File: backend/app/api/v1/health.py
from fastapi import APIRouter

from app.config import settings
from app.services.soroban import check_rpc_connectivity

router = APIRouter()


@router.get("/health")
async def health_check() -> dict[str, str]:
    """Health check endpoint.

    Probes the Soroban RPC endpoint on every call so monitoring and alerting
    see the real status of the dependency the API relies on, rather than an
    unconditional "connected".

    Also reports the network passphrase transactions are built and verified
    against. The frontend needs it to notice that a connected wallet is signing
    for a different network than this deployment anchors to — a mismatch that
    produces a signature the contract rejects, with nothing in the error to say
    why. It is reported from here rather than configured in the frontend because
    the transaction is built server-side: the backend's passphrase is the only
    thing that decides which signature will be accepted, so a second copy could
    only ever disagree with it.
    """
    rpc_connected = await check_rpc_connectivity()

    return {
        "status": "ok",
        "soroban_rpc": "connected" if rpc_connected else "unreachable",
        "network_passphrase": settings.soroban_network_passphrase,
    }

File: frontend/src/api/health.ts
import client from "./client";

/**
 * The network passphrase the backend builds and verifies transactions against.
 *
 * Read from `GET /health` rather than configured in the frontend. The anchor
 * transaction is built server-side, so the backend's passphrase is the only
 * thing that decides which signature will be accepted; a frontend copy (an env
 * var, say) could disagree with it and the disagreement would be invisible
 * until a submission was rejected without explanation.
 *
 * A failure resolves to `null`, which callers treat as "unknown" and stay quiet
 * about. Warning on a guess would be worse than not warning: an unreachable
 * health endpoint says nothing about the wallet.
 */
let cached: Promise<string | null> | null = null;

export function getContractNetworkPassphrase(): Promise<string | null> {
  cached ??= client
    .get("/health")
    .then((response) => {
      const passphrase = response.data?.network_passphrase;
      return typeof passphrase === "string" ? passphrase : null;
    })
    .catch(() => null);

  return cached;
}

/** Test seam: forget the cached answer so the next call fetches again. */
export function resetContractNetworkPassphrase(): void {
  cached = null;
}
